File: python/n110_BalanceTree.py
# https://leetcode-cn.com/problems/balanced-binary-tree/
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        if root is None: return True
        if self.isBalanced(root.left) and self.isBalanced(root.right) and abs(self.treeHeight(root.left) - self.treeHeight(root.right)) <= 1:
            return True
        else:
            return False
        
    def treeHeight(self, root: TreeNode):
        if root is None: return 0
        lev_queue = []
        lev_queue.append(root)
        level = 0
        while lev_queue:
            level += 1
            next_queue = []
            while lev_queue:              
                cur_node = lev_queue.pop(0)               
                if cur_node.left is not None: next_queue.append(cur_node.left)
                if cur_node.right is not None: next_queue.append(cur_node.right)
            lev_queue = next_queue
        return level            

File: cpp/n110_BalanceTree.cpp
// https://leetcode-cn.com/problems/balanced-binary-tree/
#include <iostream>
#include <vector>
#include <stack>
#include <algorithm>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if (root == NULL) return true;
        if (isBalanced(root -> left) && isBalanced(root -> right)) return (abs(maxDeep(root -> left) - maxDeep(root -> right)) < 2);
        else return false;
    }
    
    int maxDeep(TreeNode* root){
        if (root == NULL) return 0;
        stack <pair<TreeNode*, int>> s;
        TreeNode* p = root;
        int max_deep = 0;
        int deep = 0;
        while(!s.empty() || p != NULL){
            while(p != NULL){
                s.push(pair<TreeNode*, int>(p, ++deep));
                p = p -> left;
            }
            p = s.top().first;
            deep = s.top().second;
            if (max_deep < deep) max_deep = deep;
            s.pop();
            p = p -> right;
        }
        return max_deep;
    }
};

int main(){
    return 0;
}